maze-consulting/SFCC_Klaviyo_Maze
# Feature/kscu 7 cart rebuilding link

## Description

This update adds the capability to create a Cart Rebuilding link when the Started Checkout event occurs, and contains a new controller to handle rebuilding the cart with values passed via an encoded URL. 

Important to note _this update does not pass session information to the cart rebuilding link_ - only products, product quantity & selected options for products that were in a customer's cart. Basic error handling has been included as part of this update. 

Jira Ticket:

[KSCU-7:](https://themazegroup.atlassian.net/browse/KSCU-7) Add Cart Rebuilding link to Started Checkout event.


## Manual Testing Steps

This has been manually tested in across SFRA & SiteGen in various browsers to confirm consistency. Multiple products have been tested during development, including:

- Standard Products without special options _(ex: clothing, bags, etc.)_
- Standard Products with special options _(ex: Televisions, GPS, etc.)_
- Product Bundles w & w/o special options selected _(ex: gaming consoles, etc.)_
- Full Product Sets _(ex: winter outfits, spring outfits, etc.)_

Some test links with pre-populated products been have been added below - these are best tested in fresh incognito windows:

- [SiteGen: 6 products](https://zzqk-005.dx.commercecloud.salesforce.com/on/demandware.store/Sites-SiteGenesis-Site/default/KlaviyoRecreate-Cart?items=W3sicHJvZHVjdElEIjoiNzQwMzU3MzU4MDk1IiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOltdfSx7InByb2R1Y3RJRCI6IjcwMTY0Mjg2NzEwNCIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbXX0seyJwcm9kdWN0SUQiOiI3MDE2NDI4NTM2OTUiLCJxdWFudGl0eSI6MSwib3B0aW9ucyI6W119LHsicHJvZHVjdElEIjoibWljcm9zb2Z0LXhib3gzNjAtYnVuZGxlIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOlt7Im9wdGlvbklEIjoiY29uc29sZVdhcnJhbnR5Iiwib3B0aW9uVmFsdWVJRCI6IjAwMiIsImxpbmVJdGVtVGV4dCI6IkV4dGVuZGVkIFdhcnJhbnR5OiAzIFllYXIgV2FycmFudHkifV19LHsicHJvZHVjdElEIjoiYXBwbGUtaXBvZC10b3VjaDMyZ2IiLCJxdWFudGl0eSI6MSwib3B0aW9ucyI6W119LHsicHJvZHVjdElEIjoiNzQ5NzQzMTAtMSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbXX1d) - prod. set, prod. w/ selected option, prod. w/ selected option, prod. w/o  option
- [SiteGen: 2 products](https://zzqk-005.dx.commercecloud.salesforce.com/on/demandware.store/Sites-SiteGenesis-Site/default/KlaviyoRecreate-Cart?items=W3sicHJvZHVjdElEIjoiMDI5NDA3MzMxMjI3IiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOltdfSx7InByb2R1Y3RJRCI6Im1pY3Jvc29mdC14Ym94MzYwLWJ1bmRsZSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbeyJvcHRpb25JRCI6ImNvbnNvbGVXYXJyYW50eSIsIm9wdGlvblZhbHVlSUQiOiIwMDAiLCJsaW5lSXRlbVRleHQiOiJFeHRlbmRlZCBXYXJyYW50eTogTm9uZSJ9XX1d) - one product bundle w/o extra option and one product w/o option
- [SFRA: 14 products](https://zzys-003.dx.commercecloud.salesforce.com/on/demandware.store/Sites-RefArch-Site/default/KlaviyoRecreate-Cart?items=W3sicHJvZHVjdElEIjoic29ueS1wczMtYnVuZGxlTSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbeyJvcHRpb25JRCI6ImNvbnNvbGVXYXJyYW50eSIsIm9wdGlvblZhbHVlSUQiOiIwMDAiLCJsaW5lSXRlbVRleHQiOiJFeHRlbmRlZCBXYXJyYW50eTogTm9uZSJ9XX0seyJwcm9kdWN0SUQiOiJ2aXppby12bzM3bE0iLCJxdWFudGl0eSI6MSwib3B0aW9ucyI6W3sib3B0aW9uSUQiOiJ0dldhcnJhbnR5Iiwib3B0aW9uVmFsdWVJRCI6IjAwMyIsImxpbmVJdGVtVGV4dCI6IkV4dGVuZGVkIFdhcnJhbnR5OiA1IFllYXIgV2FycmFudHkifV19LHsicHJvZHVjdElEIjoiNzAxNjQ0MDU5MjQ4TSIsInF1YW50aXR5IjoyLCJvcHRpb25zIjpbXX0seyJwcm9kdWN0SUQiOiI3MDE2NDI4MTk0NDhNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOltdfSx7InByb2R1Y3RJRCI6IjAxMzc0MjMzNTQyMk0iLCJxdWFudGl0eSI6MSwib3B0aW9ucyI6W119LHsicHJvZHVjdElEIjoiNzQwMzU3Mzc3MDM0TSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbXX0seyJwcm9kdWN0SUQiOiI2OTMwOTI4NE0tMSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbXX0seyJwcm9kdWN0SUQiOiI4ODMzNjA1NDQyMTJNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOltdfSx7InByb2R1Y3RJRCI6IlAwMTM4TSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbXX0seyJwcm9kdWN0SUQiOiJtaWNyb3NvZnQteGJveDM2MC1idW5kbGVNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOlt7Im9wdGlvbklEIjoiY29uc29sZVdhcnJhbnR5Iiwib3B0aW9uVmFsdWVJRCI6IjAwMCIsImxpbmVJdGVtVGV4dCI6IkV4dGVuZGVkIFdhcnJhbnR5OiBOb25lIn1dfSx7InByb2R1Y3RJRCI6Im1pdHN1YmlzaGktbHQtNDAxNDhNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOlt7Im9wdGlvbklEIjoidHZXYXJyYW50eSIsIm9wdGlvblZhbHVlSUQiOiIwMDMiLCJsaW5lSXRlbVRleHQiOiJFeHRlbmRlZCBXYXJyYW50eTogNSBZZWFyIFdhcnJhbnR5In1dfSx7InByb2R1Y3RJRCI6Imdhcm1pbi1udXZpLTUwMDBNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOlt7Im9wdGlvbklEIjoiZ3BzV2FycmFudHkiLCJvcHRpb25WYWx1ZUlEIjoiMDAyIiwibGluZUl0ZW1UZXh0IjoiRXh0ZW5kZWQgV2FycmFudHk6IDMgWWVhciBXYXJyYW50eSJ9XX0seyJwcm9kdWN0SUQiOiI4ODMzNjA1MjAzMDhNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOltdfV0=) - 2 prod. sets, 2 prod. bundles, 3 prods. w/ selected option, 1 prod. w/o option
- [SFRA: 2 products](https://zzys-003.dx.commercecloud.salesforce.com/on/demandware.store/Sites-RefArch-Site/en_US/KlaviyoRecreate-Cart?items=W3sicHJvZHVjdElEIjoic29ueS1wczMtYnVuZGxlTSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbeyJvcHRpb25JRCI6ImNvbnNvbGVXYXJyYW50eSIsIm9wdGlvblZhbHVlSUQiOiIwMDAiLCJsaW5lSXRlbVRleHQiOiJFeHRlbmRlZCBXYXJyYW50eTogTm9uZSJ9XX0seyJwcm9kdWN0SUQiOiJ2aXppby12bzM3bE0iLCJxdWFudGl0eSI6MSwib3B0aW9ucyI6W3sib3B0aW9uSUQiOiJ0dldhcnJhbnR5Iiwib3B0aW9uVmFsdWVJRCI6IjAwMyIsImxpbmVJdGVtVGV4dCI6IkV4dGVuZGVkIFdhcnJhbnR5OiA1IFllYXIgV2FycmFudHkifV19XQ==) - one product bundle w/o extra option and one one product w/ selected option

## OOTB SFRA & SiteGen Estimated Total Differences

During development, it was noted the estimated totals displayed in the Cart vary slightly between SFRA & SiteGen. SFRA will display a default shipping, sales tax and estimated total as well as the price of individual line items in the cart. SiteGen, however, does not include Shipping and Sales Tax when viewing the cart for the first time or when clicking checkout for the first time _(shipping and taxes are added after proceeding as a guest / logged in user and proceeding through the steps)._ 

The links to rebuild the cart replicate this and display the prices similarly - the SFRA cart includes estimated total with shipping & taxes and individual line item costs whereas the site gen site includes individual line item costs and an estimated total without the shipping and taxes _(as though the cart is viewed for the first time)_. Noting because this appears to be the expected difference between the sites. 

## Pre-Submission Checklist:

[ x ] This update successfully builds
[ x ] No console errors are displayed on the screen with these updates
[ x ] No server-side errors / DW with ease issues are displayed with these updates
[ x ] Console Logs created during development for debugging that are no longer needed have been removed
[ n/a ] ToDo comments were left as necessary for future reference
[ x ] A description of this update & references to the manual tests has been included in this PR
